addon-test-support/@ember/test-helpers/dom/type-in.js

```diff
@@ -0,0 +1,98 @@
+import { nextTickPromise } from '../-utils';
+import settled from '../settled';
+import getElement from './-get-element';
+import isFormControl from './-is-form-control';
+import { __focus__ } from './focus';
+import { Promise } from 'rsvp';
+import fireEvent from './fire-event';
+
+/**
+ * Mimics character by character entry into the target `input` or `textarea` element.
+ *
+ * Allows for simulation of slow entry by passing an optional millisecond delay
+ * between key events.
+
+ * The major difference between `typeIn` and `fillIn` is that `typeIn` triggers
+ * keyboard events as well as `input` and `change`.
+ * Typically this looks like `focus` -> `focusin` -> `keydown` -> `keypress` -> `keyup` -> `input` -> `change`
+ * per character of the passed text (this may vary on some browsers).
+ *
+ * @public
+ * @param {string|Element} target the element or selector to enter text into
+ * @param {string} text the test to fill the element with
+ * @param {number} millisecondDelay number of millisecons to wait per keypress
+ * @return {Promise<void>} resolves when the application is settled
+ */
+export default function typeIn(target, text, millisecondDelay = 0) {
+  return nextTickPromise().then(() => {
+    if (!target) {
+      throw new Error('Must pass an element or selector to `typeIn`.');
+    }
+
+    const element = getElement(target);
+    if (!element) {
+      throw new Error(`Element not found when calling \`typeIn('${target}')\``);
+    }
+    let isControl = isFormControl(element);
+    if (!isControl) {
+      throw new Error('`typeIn` is only usable on form controls.');
+    }
+
+    if (typeof text === 'undefined' || text === null) {
+      throw new Error('Must provide `text` when calling `typeIn`.');
+    }
+
+    __focus__(element);
+
+    return fillOut(element, text, millisecondDelay)
+      .then(() => fireEvent(element, 'change'))
+      .then(settled);
+  });
+}
+
+// eslint-disable-next-line require-jsdoc
+function fillOut(element, text, delay) {
+  return new Promise(resolve => {
```

Comment:
I don't understand this promise nesting

Reply:
The reason for the promise nesting is because `fillOut` takes each step and executes `keyDown, keyPress, input and keyUp` events for each character in the input string.  

We need to ensure that each promise is triggered in order.

Without this, everything essentially happens without delay.

Reply:
Yep, your right.  Have removed.  Sorry for misunderstanding the original question!